src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala

```diff
@@ -205,6 +206,21 @@ private class TaskLauncherActor(
   }
 
   private[this] def receiveTaskLaunchNotification: Receive = {
+    case InstanceOpSourceDelegate.InstanceOpRejected(op, TaskLauncherActor.OfferOperationRejectedTimeoutReason) =>
+      // Reschedule instance with provision timeout..
+      if (inFlightInstanceOperations.exists(_.instanceId == op.instanceId)) {
+        instanceMap.get(op.instanceId).foreach { instance =>
+
+          /**
+            * We don't await the future. It will trigger an instance update event.
+            *
+            * The rescheduled instance forgets about all tasks. If these tasks do come back in an update they a killed.
+            * See [[mesosphere.marathon.core.task.update.impl.TaskStatusUpdateProcessorImpl.publish()]] for the logic.
+            */
+          instanceTracker.schedule(instance.rescheduled())
```

Comment:
The fact that we write to this based on our cached copy without any consideration for modifications made by other modules makes me pretty nervous.

Reply:
yep, there were few races around that. At the same time, we receive all InstanceChange events through event stream.

On the other hand, if we process this timeout and at the same time someone persists some change to the instance, we then can be acting on stale state, true...